### ISSUE TITLE
Support for bun runtime

### ISSUE THREAD

**Issue by asilvas** [lovell/sharp#3511]
### What are you trying to achieve?

[Bun](https://bun.sh/) seems like a good fit for sharp/vips with the focus on performance. The reduction in overhead from bindings could be a win with Bun. Currently no good image processing package for Bun, so this would given even more reason to promote sharp.

### When you searched for similar feature requests, what did you find that might be related?

Nothing

### What would you expect the API to look like?

n/a

### What alternatives have you considered?

Currently use sharp extensively with Node.

### Please provide sample image(s) that help explain this feature

n/a

```
[8.68ms] "node_modules.bun" - 15 modules, 8 packages
29 |     if (loadedModule) {
30 |       const [, loadedPackage] = loadedModule.match(/node_modules[\\/]([^\\/]+)[\\/]/);
31 |       help.push(`- Ensure the version of sharp aligns with the ${loadedPackage} package: "npm ls sharp"`);
32 |     }
33 |   }
34 |   throw new Error(help.join('\n'));
            ^
error:
Something went wrong installing the "sharp" module

Cannot find module "../build/Release/sharp-linux-x64.node" from "./node_modules/sharp/lib/sharp.js"

Possible solutions:
- Install with verbose logging and look for errors: "npm install --ignore-scripts=false --foreground-scripts --verbose sharp"
- Install for the current linux-x64 runtime: "npm install --platform=linux --arch=x64 sharp"
- Consult the installation documentation: https://sharp.pixelplumbing.com/install
```

**Comment by lovell** [comment]
This will be dependent on https://github.com/oven-sh/bun/issues/158 and https://github.com/oven-sh/bun/issues/606

**Comment by asilvas** [comment]
> This will be dependent on [oven-sh/bun#158](https://github.com/oven-sh/bun/issues/158) and [oven-sh/bun#606](https://github.com/oven-sh/bun/issues/606)

Great thanks. I suspect those issues will be unblocked in the next couple months. I'll try to remember to update this thread once ready.

**Comment by birkskyum** [comment]
@lovell , would you be able to narrow down what parts of https://github.com/oven-sh/bun/issues/158 are lacking in order for sharp to function?

**Comment by birkskyum** [comment]
Okay, interesting that it could be resolved already. The other issue mentioned breaks on installation (https://github.com/oven-sh/bun/issues/606), so let's assess the situation when that is fixed.

**Comment by JawadAhbab** [comment]
Can anyone give me an working example please?

**Comment by paperdave** [comment]
Once Bun 0.8.0 releases, Sharp will work (napi features sharp depends on is more stable/fixed a bad crash), though the NPM lifecycle script not running is still an issue, see https://github.com/oven-sh/bun/issues/606

**Comment by xHyroM** [comment]
For now, you can manually run post install script

**Comment by birkskyum** [comment]
@xHyroM , how do i run `postinstall` with bun?

**Comment by xHyroM** [comment]
> @xHyroM , how do i run `postinstall` with bun?

```sh
cd node_modules/sharp
bun install
```

then wait until install and try to run your code again, it should work :D

**Comment by birkskyum** [comment]
@xHyroM , it works directly on my machine, which is awesome, but when doing this within a Dockerfile I get:

```sh
error: script not found "install/libvips"
error: script not found "install/can-compile"
```

My entire dockerfile is:

```dockerfile
FROM oven/bun

WORKDIR /usr/repo

COPY . .

RUN bun install
RUN cd ./node_modules/sharp && bun install

EXPOSE 3001

CMD [ "bun", "--bun", "run", "./src/app.ts" ]

```

**Comment by birkskyum** [comment]
Seems like the script can be found by making the paths in sharp's package.json relative, so they are i.e. `./install/libvips.js` install of `install/libvips.js`. Will open a PR on that.

**Comment by xHyroM** [comment]
Not sure what to do, with npm you could do `npm rebuild --arch=x64 --platform=linux sharp` but you want bun

**Comment by birkskyum** [comment]
I realized that the bun runtime works with sharp, but that `bun install` doesn't, so an easy solution was to just install with npm and run with bun after.

**Comment by birkskyum** [comment]
This will likely be resolved by:
- https://github.com/oven-sh/bun/pull/4263

**Comment by xHyroM** [comment]
You can try to put sharp into trusped deps and try bun install again, it should run the install script
^ want to know if it'll resolve the issue you send above

**Comment by xHyroM** [comment]
> Normal Sharp install in Bun will likely be resolved by:
> 
> * [feat(install): include a default `trustedDependencies` allowlist oven-sh/bun#4263](https://github.com/oven-sh/bun/pull/4263)
> 
> Then sharp just has to be added in package.json as `trustedDependencies: ["sharp"]`

Can you try to put sharp into `trustedDependencies` and try bun install again? It should run the install script
^ want to know if it'll resolve the issue you send above

**Comment by birkskyum** [comment]
With latest canary, and sharp in trustedDependencies, I get a:

```
➜ bun run start
$ bun run ./src/app.ts
[0.01ms] ".env"
error: script "start" exited with code 11 (SIGSEGV)
```

**Comment by infrahead** [comment]
Simply using `npm install` in case of sharp seems to be the answer for now.  I'd also suggest that the real solution will be #3750 and not Bun's trustedDependencies convenience feature, as some will never allow install scripts.

**Comment by birkskyum** [comment]
@lovell this can be closed since the change required is not to sharp, but rather to bun install

**Comment by birkskyum** [comment]
I made a separate ticket for supporting bun install. This ticket about bun runtime appear to be resolved.

**Comment by Jarred-Sumner** [comment]
> With latest canary, and sharp in trustedDependencies, I get this, and i think it's because it can't find the arm64 bundle:
> 
> ```
> ➜ bun run start
> $ bun run ./src/app.ts
> [0.01ms] ".env"
> error: script "start" exited with code 11 (SIGSEGV)
> ```

It is unlikely that bun not finding the .node file would cause a segfault. It sounds like we still have a crash to fix here

**Comment by birkskyum** [comment]
Running with the debug build it looks like this:

```zig
loader [parsing] /Users/admin/repos/backend/node_modules/stripe/esm/autoPagination.js
Loader [resolve] ./utils.js
loader [parsing] /Users/admin/repos/backend/node_modules/stripe/esm/multipart.js
Loader [resolve] ./utils.js
32 |     if (loadedModule) {
33 |       const [, loadedPackage] = loadedModule.match(/node_modules[\\/]([^\\/]+)[\\/]/);
34 |       help.push(`- Ensure the version of sharp aligns with the ${loadedPackage} package: "npm ls sharp"`);
35 |     }
36 |   }
37 |   throw new Error(help.join('\n'));
            ^
error: 
Something went wrong installing the "sharp" module

Cannot find module "../build/Release/sharp-darwin-arm64v8.node" from "/Users/admin/repos/backend/node_modules/sharp/lib/sharp.js"

Possible solutions:
- Install with verbose logging and look for errors: "npm install --ignore-scripts=false --foreground-scripts --verbose sharp"
- Install for the current darwin-arm64v8 runtime: "npm install --platform=darwin --arch=arm64v8 sharp"
- Consult the installation documentation: https://sharp.pixelplumbing.com/install
      at /Users/admin/repos/backend/node_modules/sharp/lib/sharp.js:37:8
      at globalThis (/Users/admin/repos/backend/node_modules/sharp/lib/sharp.js:37:33)
```

**Comment by qorbani** [comment]
@hendrikboeck in your message and code snippet you misspelled `trustedDependencies` 😄

**Comment by lovell** [comment]
Please take a look at https://github.com/lovell/sharp/issues/3750#issuecomment-1751172728 if you'd like to test an early pre-release of the forthcoming sharp v0.33.0 that should "just work" with Bun without setting `trustedDependencies`.

**Comment by MariuzM** [comment]
Sharp works locally with Bun but when i deploy to railway i get this error 

<img width="805" alt="image" src="https://github.com/lovell/sharp/assets/36937039/52cca1c6-a71f-43cd-85af-7dea98e372e0">

Only diff is that i'm using docker on railway 

```docker
FROM oven/bun:1.0.12-alpine

WORKDIR /usr/src/app
COPY . .
RUN bun install
RUN bun run build

ENV PORT 3001

EXPOSE 3000

CMD bun run start
```

**Comment by lovell** [comment]
@MariuzM https://github.com/lovell/sharp/pull/3848 will fix the wrong error being shown, but there's still an underlying problem. It looks like you're using some kind of bundler, so please ensure sharp is marked as "external" - see https://sharp.pixelplumbing.com/install#bundlers

**Comment by MariuzM** [comment]
I don't think i use anything else apart what Bun already gives, my build command is 
`bun build ./src/app.ts --outdir ./build --target bun` 

i have tried 
`bun build ./src/app.ts --outdir ./build --target bun --external:sharp` 

but still same issue

**Comment by lovell** [comment]
@MariuzM This looks like something specific to the `bun build` command itself. Perhaps you could create a minimal, standalone repo that allows someone else to reproduce.

**Comment by MariuzM** [comment]
@lovell https://github.com/MariuzM/bun-sharp and i have tested on Railway still same issue

**Comment by MariuzM** [comment]
Ive just tested locally with docker and issue is there, and to confirm without docker locally i dont have this issue

But i don't know what the issue might be, docker config looks simple enough 

<img width="832" alt="image" src="https://github.com/lovell/sharp/assets/36937039/39ee0c18-58a4-4c23-a2cb-3f4adf6fc0b6">

```docker
# https://hub.docker.com/r/oven/bun/tags?page=1&name=alpine
FROM oven/bun:1.0.12-alpine

WORKDIR /usr/src/app
COPY . .
RUN bun install
RUN bun run build

ENV PORT 3001

EXPOSE 3000

CMD bun run start
```

**Comment by MariuzM** [comment]
Ive just tested locally with Docker and issue is there, but when NOT using docker i don't have this issue, so not sure what difference does docker make, my docker config looks simple enough 

```
# https://hub.docker.com/r/oven/bun/tags?page=1&name=alpine
FROM oven/bun:1.0.12-alpine

WORKDIR /usr/src/app
COPY . .
RUN bun install
RUN bun run build

ENV PORT 3001

EXPOSE 3000

CMD bun run start
```

<img width="836" alt="image" src="https://github.com/lovell/sharp/assets/36937039/df7b85a5-57eb-453b-82d6-32d88142cc25">

**Comment by hendrikboeck** [comment]
@MariuzM I found a solution that worked for me in Docker v24.0.7:

**Comment by hendrikboeck** [comment]
@MariuzM I fixed your problem. I think you forgot to add `semver` to your `package.json` and forgot to install `vips-dev` in Alpine, as you declared it as external.

my Dockerfile:
```Dockerfile
# https://hub.docker.com/r/oven/bun/tags?page=1&name=alpine
FROM oven/bun:1.0.12-alpine

RUN apk add vips-dev

WORKDIR /usr/src/app
COPY . .
RUN bun install
RUN bun run build

EXPOSE 3000
CMD bun run start
```

my package.json:
```json
{
  "name": "bun-sharp",
  "version": "1.0.0",
  "scripts": {
    "d": "bun run --watch src/app.ts",
    "b": "bun build ./src/app.ts --outdir ./build --target bun --external sharp",
    "dev": "bun run d",
    "build": "bun run b",
    "start": "bun run ./build/app.js"
  },
  "dependencies": {
    "elysia": "^0.7.28",
    "semver": "^7.5.4",
    "sharp": "0.33.0-alpha.11"
  },
  "devDependencies": {
    "@types/node": "^20.9.1",
    "@types/uuid": "^9.0.7",
    "@typescript-eslint/eslint-plugin": "^6.11.0",
    "@typescript-eslint/parser": "^6.11.0",
    "bun-types": "^1.0.12",
    "eslint": "8.54.0",
    "eslint-plugin-import": "^2.29.0",
    "prettier": "^3.1.0"
  },
  "module": "src/app.js"
}
```

my .dockerignore:
```
/node_modules
/build
```

**Comment by MariuzM** [comment]
@hendrikboeck thank you this worked. Its not that i forgot i just did not know 😁

**Comment by lovell** [comment]
@MariuzM Whilst it might appear to work, there is no need to add the `vips-dev` package. In addition, the fact that you also had to add a dependency on `semver` despite your code not depending upon it does rather suggest that `bun build` is still not working as advertised.

**Comment by lovell** [comment]
@MariuzM Your container uses Alpine Linux but Bun does not officially support musl-based Linux - https://github.com/oven-sh/bun/issues/918

In addition, Bun polyfills the `detect-libc` dependency and returns a hard-coded `glibc` on all Linux platforms - see https://github.com/oven-sh/bun/blob/main/src/js/thirdparty/detect-libc.linux.js

**Comment by MariuzM** [comment]
@lovell thank you for taking time to explain in more detail, dint realized Alpine had missing packages.

**Comment by lovell** [comment]
v0.33.0 is now available and tested working with the Bun JavaScript runtime on both Linux and macOS.

https://github.com/lovell/sharp/actions/runs/7032565186

**Comment by MariuzM** [comment]
THANK YOUUUUU

**Comment by lovell** [comment]
@7heMech It looks like you're using `0.33.0-alpha.6`, which is an unsupported pre-release version. Please also ensure you're using the latest Bun as the `.replit` file suggests an older version.
